src/blockMonitor/reorgEmitter.ts

```diff
@@ -0,0 +1,198 @@
+import { ethers } from "ethers";
+import { StartStopService, ApplicationError } from "../dataEntities";
+import { IBlockStub } from "./blockStub";
+import { ReorgHeightListenerStore } from "./reorgHeightListener";
+import { BlockProcessor } from "./blockProcessor";
+import { Lock } from "../utils/lock";
+
+/**
+ * Keeps track of the current head of the blockchain, and handles reorgs up to the same depth maxDepth as the
+ * BlockCache that is passed in the constructor.
+ * Emits appropriate events when reorgs are observed, and resets the provider to the common ancestor (whenever possible)
+ * so that appropriate block events are emitted.
+ */
+export class ReorgEmitter extends StartStopService {
+    private headBlock: IBlockStub;
```

Comment:
Probably true, and I changed the `handleReorgEvent` to use `oldHeadHash` instead. On the other hand, having an explicitly set `head` in the reorgEmitter is heavily used in the unit tests.

Meanwhile, I realized that I wasn't correctly type safe in the headBlock field (could be null sometimes); improved it in 6057e89.

Reply:
KK, if it's used heavily in the tests then lets keep it for now. But we should remove it - and refactor the tests in the long term, since I think they can probably done without out this method. I've added a note to #187